### ISSUE TITLE
AlertManager, ruler, config-Api - how to configure them and get it working.

### ISSUE THREAD

**Issue by mizunos** [cortexproject/cortex#4050]
Really appreciate some helps as I am struggling with figuring out how to use cortex alert manager and ruler.  This is my last resort as I posted on Slack but got no answer for a while and I am hoping to get some guidance as we are blocked to go production if we don't have any way to get alert on our devices.  The documentation is quite confusing and if you guys help me with working through this,  I offer my help to document the process to implement and operate this important area so that we don't have to go pour though the code in order to figure out how this work.  I would pour through the code to figure this out but I am pressed for time at the moment.  

I deployed alertmanager, ruler, config and config-db in GKE(version 1.4).  All pods seemed to be running.  I can port-forward to alertmanager, ruler, configs  pods to look at the configuration.  I am using block configuration for collecting the data but using the configdb for alertmanager and ruler as well as nginx reverse proxy but I cannot seem to reach alertmanager or ruler externally so not sure how to perform CRUD on rules.  Getting 404 error on different attempt to get to alertmanager

Do I need to set alert rules as a configmap as in Prometheus alert or using HTTP API?. 
Do I used config-API to setup the rules and if so where to set that up?
Any suggestion is greatly appreciated.

Here are the code snippets for all of them:

nginx configmap

```
http {
      default_type application/octet-stream;
      log_format   main '$remote_addr - $remote_user [$time_local]  $status '
        '"$request" $body_bytes_sent "$http_referer" '
        '"$http_user_agent" "$http_x_forwarded_for"';
      access_log   /dev/stderr  main;
      sendfile     on;
      tcp_nopush   on;
      resolver kube-dns.kube-system.svc.cluster.local;

      server { # simple reverse-proxy
        listen 80;
        proxy_set_header X-Scope-OrgID 0;

        location = /api/v1/push {
          proxy_pass      http://distributor.monitoring.svc.cluster.local$request_uri;
        }

        location = /api/prom/push {
          proxy_pass      http://distributor.monitoring.svc.cluster.local$request_uri;
        }

        location = /ring {
          proxy_pass      http://distributor.monitoring.svc.cluster.local$request_uri;
        }
        location = /all_user_stats {
          proxy_pass      http://distributor.monitoring.svc.cluster.local$request_uri;
        }

        location ~ /prometheus/.* {
          proxy_pass      http://query-frontend.monitoring.svc.cluster.local$request_uri;
        }

        location ~ /api/prom/.* {
          proxy_pass      http://query-frontend.monitoring.svc.cluster.local$request_uri;
        }
        location = /alertmanager {
          proxy_pass      http://alertmanager.monitoring.svc.cluster.local$request_uri;
        }
        location = /configs {
          proxy_pass      http://configs.monitoring.svc.cluster.local$request_uri;
        }
```

alertmanager deployment snippet
```
      containers:
      - name: alertmanager
        image: quay.io/cortexproject/cortex:v1.4.0
        imagePullPolicy: IfNotPresent
        args:
        - -target=alertmanager
        - -log.level=debug
        - -server.http-listen-port=80
        - -alertmanager.storage.retention=240h
        - -alertmanager.configs.poll-interval=10s
        - -alertmanager.configs.url=http://configs.monitoring.svc.cluster.local:80
        - -alertmanager.web.external-url=/api/prom/alertmanager/
        - -alertmanager.storage.path=/data
        - -experimental.alertmanager.enable-api=true

        ports:
        - containerPort: 80
```
When port-forward to alertmanager pod here is what I get

**Service status
-- | --
-- server | Running
-- alertmanager | Running

ruler deployment code snippet
```
        args:
        - -target=ruler
        - -log.level=debug
        - -store.engine=blocks
        - -server.http-listen-port=80
        - -ruler.configs.url=http://configs.monitoring.svc.cluster.local:80
        - -ruler.alertmanager-url=http://alertmanager.monitoring.svc.cluster.local/api/prom/alertmanager/
        - -ruler.ring.store="consul"
        - -ruler.alertmanager-use-v2
        - -consul.hostname=consul.monitoring.svc.cluster.local:8500
        - -distributor.replication-factor=2
        - -querier.store-gateway-addresses=store-gateway.monitoring.svc.cluster.local:9095
        - -blocks-storage.backend=gcs
        - -blocks-storage.gcs.bucket-name=edge-monitor-block-storage
        - -blocks-storage.bucket-store.sync-interval=5m
        - -blocks-storage.bucket-store.max-concurrent=100
        - -blocks-storage.bucket-store.index-cache.backend=memcached
        - -blocks-storage.bucket-store.index-cache.memcached.addresses=memcached.monitoring.svc.cluster.local:11211
        - -blocks-storage.bucket-store.index-cache.postings-compression-enabled=true
        - -blocks-storage.bucket-store.chunks-cache.backend=memcached
        - -blocks-storage.bucket-store.chunks-cache.memcached.addresses=memcached.monitoring.svc.cluster.local:11211
        - -blocks-storage.bucket-store.metadata-cache.backend=memcached
        - -blocks-storage.bucket-store.metadata-cache.memcached.addresses=memcached.monitoring.svc.cluster.local:11211
        ports:
        - containerPort: 80
        volumeMounts:
        - name: config-volume
          mountPath: /etc/cortex
        - name: secrets
          mountPath: /etc/secret
          readOnly: true
      volumes:
        - name: config-volume
          configMap:
            name: schema-config
#new
        - name: secrets
          secret:
            secretName: storage-key
            items:
            - key: storage.json
              path: storage.json
```

Log snippet of ruler:

```
level=debug ts=2021-04-06T15:49:14.431387671Z caller=module_service.go:48 msg="module waiting for initialization" module=distributor-service waiting_for=ring                                                                   │
│ level=info ts=2021-04-06T15:49:14.432050894Z caller=module_service.go:58 msg=initialising module=server                                                                                                                         │
│ level=debug ts=2021-04-06T15:49:14.517584859Z caller=module_service.go:48 msg="module waiting for initialization" module=ring waiting_for=memberlist-kv                                                                         │
│ level=info ts=2021-04-06T15:49:14.517641696Z caller=module_service.go:58 msg=initialising module=ring                                                                                                                           │
│ level=info ts=2021-04-06T15:49:14.533579286Z caller=module_service.go:58 msg=initialising module=distributor-service                                                                                                            │
│ level=debug ts=2021-04-06T15:49:14.53501856Z caller=module_service.go:48 msg="module waiting for initialization" module=ruler waiting_for=store-queryable                                                                       │
│ level=debug ts=2021-04-06T15:49:38.174471248Z caller=logging.go:66 traceID=a4e7f7b2ec1fb55 msg="GET /metrics (200) 33.795685ms"                                                                                                 │
│ level=info ts=2021-04-06T15:49:54.309294523Z caller=fetcher.go:453 org_id=0 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=39.752934351s cached=1805 returned=1805 partial=0               │
│ level=debug ts=2021-04-06T15:50:08.146712657Z caller=logging.go:66 traceID=20a3fdaa61db823e msg="GET /metrics (200) 7.320934ms"                                                                                                 │
│ level=info ts=2021-04-06T15:50:29.809331644Z caller=module_service.go:58 msg=initialising module=ruler                                                                                                                          │
│ level=info ts=2021-04-06T15:50:29.809518979Z caller=ruler.go:345 msg="ruler up and running"                                                                                                                                     │
│ level=info ts=2021-04-06T15:50:29.834176675Z caller=cortex.go:341 msg="Cortex started"                                                                                                                                          │
│ level=debug ts=2021-04-06T15:50:38.146054155Z caller=logging.go:66 traceID=6df1afc0214c63f9 msg="GET /metrics (200) 6.793618ms"                                                                                                 │
│ level=debug ts=2021-04-06T15:51:08.146738151Z caller=logging.go:66 traceID=3bb35ec2f22c1a89 msg="GET /metrics (200) 7.516703ms"                                                                                                 │
│ level=debug ts=2021-04-06T15:51:38.146142407Z caller=logging.go:66 traceID=4b681f1d80e30105 msg="GET /metrics (200) 6.926123ms"                                                                                                 │
│ level=debug ts=2021-04-06T15:51:55.751703646Z caller=logging.go:66 traceID=44bab603e25b2523 msg="GET / (200) 29.211969ms"                                                                                                       │
│ level=debug ts=2021-04-06T15:51:55.956733846Z caller=logging.go:66 traceID=c44c7c9fd5f31e msg="GET /favicon.ico (404) 89.821µs"                                                                                                 │
│ level=debug ts=2021-04-06T15:52:03.5442551Z caller=logging.go:66 traceID=1ca581543b68cd8e msg="GET /ingester/ring (200) 1.021307ms"                                                                                             │
│ level=debug ts=2021-04-06T15:52:07.547858085Z caller=logging.go:66 traceID=2c974d1714951c84 msg="GET /ruler/ring (200) 81.974µs"                                                                                                │
│ level=debug ts=2021-04-06T15:52:08.146473993Z caller=logging.go:66 traceID=7a84671dc82cb0c4 msg="GET /metrics (200) 7.3642ms"                                                                                                   │
│ level=debug ts=2021-04-06T15:52:12.48044548Z caller=logging.go:66 traceID=7ca0aab62a75abf4 msg="GET /services (200) 105.629662ms"                                                                                               │
│ level=debug ts=2021-04-06T15:52:38.146183129Z caller=logging.go:66 traceID=2732cabbed71dcdb msg="GET /metrics (200) 7.169771ms"
```

**Comment by Carmezim** [comment]
It'd be helpful indeed to have more documentation on configuring alerts for cortex 🙏

**Comment by mizunos** [comment]
We can close this issue as I stumble through and manage to upload rules,  alertmanager configuration for notification and will start doing some test.  This is probably only 1/3 of the road for us to be able to fully used all the features and alert manager. For example,  just encounter some other error on alertmanager configuration that I could not find any information.  More than happy to document my little crazy adventure with Cortex.  All I can say is I learn a lot stumbling through this with some great help for block storage from grafana/cortex team. I found most of the time my post on #cortex Slack channel does not generate much reply.

**Comment by Carmezim** [comment]
@mizunos if you could share more details of how you configured (assuming Config API) and the errors you've found that'd be helpful I'm sure for the cortex guys and others 🙏

**Comment by mizunos** [comment]
I am not using Config API as from reading different post,  they said that Config API is deprecated.  In short 
1 - to load rules into ruler I am using Postman,  port forwarding ruler service on K8s and use POST /api/v1/rules/{namespace}  with the body containing the Yaml for rules
2 - Load alertmanager configuration for notification,  I use curl to load the alertmanager configuration yaml file.  (For some weird reason,  the combination approach in 1 did not work and I have no time to look into it yet

While both alert rules and notfication is in,  I have not been able to full y test end to end yet because I now have other error pop up in alert manager log that I am trying to figure out.

I now have another issue opened 

 
https://github.com/cortexproject/cortex/issues/4080

**Comment by mizunos** [comment]
I have seen post of others using cortextool but I am not able to get it to work yet as the doc is a little unclear on how to set paramters to work with cortex running in K8s environment and how to set basic auth etc..

**Comment by pracucci** [comment]
> 1 - to load rules into ruler I am using Postman, port forwarding ruler service on K8s and use POST /api/v1/rules/{namespace} with the body containing the Yaml for rules
> 2 - Load alertmanager configuration for notification, I use curl to load the alertmanager configuration yaml file. (For some weird reason, the combination approach in 1 did not work and I have no time to look into it yet

I would suggest you to use https://github.com/grafana/cortex-tools#cortextool. It provides a CLI to operate on the alertmanager and ruler APIs.

**Comment by Carmezim** [comment]
@mizunos thank you for sharing!

**Comment by Carmezim** [comment]
@mizunos did you have any issues with permissions when configuring the alerts like

```
error validating Alertmanager config: mkdir /tmp/validate-config798960130: read-only file system
```

**Comment by pstibrany** [comment]
> @mizunos did you have any issues with permissions when configuring the alerts like
> 
> ```
> error validating Alertmanager config: mkdir /tmp/validate-config798960130: read-only file system
> ```

When user uploads config, it is validated first. During this validation, config is stored to a temporary directory, and this directory needs to be writeable. The error message is saying that yours isn’t.

**Comment by Carmezim** [comment]
@pstibrany thanks Peter. I responded in the ticket you replied about that too and indeed was just the helm chart hardcoding permissions requiring to change it on a fork. 

I've got another question though, I've got the alertmanager configured and the ruler as well with the recording and alerting rules (I can retrieve them through the API and the alertmanager see the config also in the UI). 

I still wasn't able to get alerts fired though. Is there any insights or first places to look at in these cases you'd recommend? 

Thank you very much 🙏
